src/NUnitFramework/framework/Internal/Builders/NUnitTestCaseBuilder.cs

```diff
@@ -75,23 +74,11 @@ public TestMethod BuildTestMethod(IMethodInfo method, Test parentSuite, TestCase
 
             if (parms != null)
             {
-                parms.ApplyToTest(testMethod);
-
-                if (parms.TestName != null)
-                {
-                    // The test is simply for efficiency
-                    testMethod.Name = parms.TestName.Contains("{")
-                        ? new TestNameGenerator(parms.TestName).GetDisplayName(testMethod, parms.OriginalArguments)
-                        : parms.TestName;
-                }
-                else
-                {
-                    testMethod.Name = _nameGenerator.GetDisplayName(testMethod, parms.OriginalArguments);
-                }
+                parms.ApplyToTest(testMethod, _nameGenerator);
             }
             else
             {
-                testMethod.Name = _nameGenerator.GetDisplayName(testMethod, null);
+                testMethod.Name = _nameGenerator.GetDisplayName(testMethod, null, null);
```

Comment:
I'm not a massive fan of how this code has ended up reading - namely how it looks like the TestName is generated entirely differently depending on whether there are params or not. 

I'd prefer this to be `_nameGenerator.GetDisplayName()` in all cases, and for TestNameGenerator to look after all the resolution of testnames, and deal with all the if/else cases which are currently scattered through TestCaseBuilder/TestFixtureBuilder/TestNameGenerator/TestParameters.

Would that be technically possible, or am I missing why it's currently done this way? Is it just that doing that would be a bigger refactor than this PR has set out to do? It feels like we're changing the structure fundamentally enough at this point for that to be worthwhile.

Reply:
I made the builder stop reaching into the parameters and pulling out pieces of data like TestName and ArgNames and possibly future properties, then making decisions based on those properties before applying the result to test.Name.

Practically speaking, it allowed me to make the ArgNames property protected instead of public and also to not duplicate the whole chunk of logic when applying the test fixture name.

Pattern-wise, it's consistent with the direction we're moving with everything that was already in the ApplyTo method. Test parameters decide how to apply the Name rather than querying TestName and now ArgNames and possibly future properties.

Reply:
Ah ok. I see the ApplyTo architecture was pre-existing. I'm not sure I'm totally a fan in this instance, where the test name is sometimes set by TestParams, and sometimes fetched from the NameGenerator, but I see when this has been done like it is. 🙂